### PR TITLE
[FLINK-29028][python][doc] Add the missing cache api in Python DataStream API

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pull request will add the missing cache api in Python DataStream API*

## Brief change log

  - *Add the missing cache api in Python DataStream API*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/pyflink/datastream/data_stream.py [hunk 1]
```diff
@@ -895,6 +895,20 @@ def get_side_output(self, output_tag: OutputTag) -> 'DataStream':
...
+        currently only block mode is supported. The cache is generated lazily at the first time the
+        intermediate result is computed. The cache will be clear when the StreamExecutionEnvironment
+        close.
+
+        :return: The cached DataStream that can use in later job to reuse the cached intermediate
+                 result.
+
+        .. versionadded:: 1.16.0
+        """
+        return DataStream(self._j_data_stream.cache())
```
Comment (dianfu) [inline flink-python/pyflink/datastream/data_stream.py:910]: I guess we may need to return CachedDataStream which has some specific API, e.g. invalidate.

File: flink-python/pyflink/datastream/data_stream.py [hunk 2]
```diff
@@ -895,6 +895,20 @@ def get_side_output(self, output_tag: OutputTag) -> 'DataStream':
...
+    def cache(self) -> 'DataStream':
+        """
+        Cache the intermediate result of the transformation. Only support bounded streams and
+        currently only block mode is supported. The cache is generated lazily at the first time the
+        intermediate result is computed. The cache will be clear when the StreamExecutionEnvironment
+        close.
+
+        :return: The cached DataStream that can use in later job to reuse the cached intermediate
+                 result.
+
```
Comment (dianfu) [inline flink-python/pyflink/datastream/data_stream.py:907]: Need also add close method in StreamExecutionEnvironment.

File: flink-python/pyflink/datastream/data_stream.py (hunk 1 above)
Comment (Sxnan) [reply flink-python/pyflink/datastream/data_stream.py:910]: On top of that, we also introduce `SideOutputDataStream` which supports the cache method. I think we may need to introduce `SideOutputDataStream` in PyFlink and let get_side_output return `SideOutputDataStream`.

Comment (HuangXingBo) [reply flink-python/pyflink/datastream/data_stream.py:910]: @Sxnan I think we can reuse the `cache` method of `DataStream`, no need to introduce a new class `SideOutputDataStream`

**Comment by HuangXingBo** [comment]
@dianfu @Sxnan I have resolved the comments in the latest commit.

**Review by dianfu — APPROVED** [review]

**Pull request closed without merging**